### ISSUE TITLE
Firing pop and then navigate leads to navigate followed by pop action

### ISSUE THREAD

**Issue by paour** [react-navigation/react-navigation#5062]
### Current Behavior

- stackNavigator inside a stackNavigator
- pop followed by navigate
- the screen is correctly popped, the navigation happens, but then an extra pop is performed and the screen that was just navigated to disappears

### Expected Behavior

- only the current screen should be popped, not the one navigated to

### How to reproduce

https://snack.expo.io/@paour/poptoomuch

First tap "Inner2", then "Outer2 No Delay": the Inner2 screen is popped, Outer2 briefly shows, but is the replaced with Inner1 after a second extraneous pop is performed.

Tapping the "Outer2 With Delay" uses a listener to only perform the navigation once the pop is performed. This works from a navigation standpoint, but the desired outcome is that the inner navigation is popped in the background while the outer navigation happens, without an intermediate animation and display of inner1.

### Your Environment

| software         | version
| ---------------- | -------
| react-navigation | 2.17.0
| react-native     | latest expo

**Comment by brentvatne** [comment]
this does indeed seem to be an issue. I think it's something that anyone can investigate and get to the bottom of, so I've tagged it with help wanted and hacktoberfest, hopefully someone can help!

**Comment by tienf4m** [comment]
I believe this happens because the pop transition ends after the navigate transition, so calling navigate with a delay wouldn't cause this problem. I don't think we should change the route index when it's a COMPLETE_TRANSITION action, simply adding COMPLETE_TRANSITION to this line below would fix the problem.
https://github.com/react-navigation/react-navigation/blob/4bd6f17b466bdbb0d0035cf0b8e779d6343dedd2/src/routers/StackRouter.js#L547
What do you think? I'm happy to make a PR :)

**Comment by brentvatne** [comment]
cc @ericvicenti 

@tienf4m - if this does indeed fix it a PR would be welcome!

**Comment by watanabeyu** [comment]
Hi.
this works for me.
```js
if (childRouter) {
          const route = childRouter.getStateForAction(action, childRoute);

          if (route === null) {
            return state;
          } else if (route && route !== childRoute) {
            console.log(action.type)
            return StateUtils.replaceAt(
              state,
              childRoute.key,
              route,
              // the following tells replaceAt to NOT change the index to this route for the setParam action, because people don't expect param-setting actions to switch the active route
              // (action.type === NavigationActions.SET_PARAMS || action.type === StackActions.COMPLETE_TRANSITION)
              (action.type === NavigationActions.SET_PARAMS || action.type === StackActions.COMPLETE_TRANSITION)
            );
          }
        }
```

I changed `action.type === NavigationActions.SET_PARAMS` -> `(action.type === NavigationActions.SET_PARAMS || action.type === StackActions.COMPLETE_TRANSITION)`.

Please anyone send PR.

**Comment by brentvatne** [comment]
I released @tienf4m's PR in 2.18.0 and you can confirm the fix by opening this fork of the original Snack that uses the latest version of react-navigation: https://snack.expo.io/@notbrent/poptoomuch

**Comment by paour** [comment]
Confirmed, thank you for this amazing turnaround!
